### PR TITLE
feature: wallet microservice

### PR DESCRIPTION
**Pull request by netpoe**
This PR lifts up a `wallet` microservice created with the `@aufacicenta/microservice-generator@1.0.7`. 

If your database engine is already running, simply run `cd wallet; source .env; yarn start` and the service should start without errors. 

Screen Shot 2021-05-31 at 15 15 54

Screen Shot 2021-05-31 at 15 16 00

### CODE REVIEW

File: wallet/src/service/controller.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    {
+      call,
+      callback,
+    }: gRPCServerUnaryCall<CreateWalletRequest, CreateWalletReply>,
+    { dao }: IContext
+  ) {
+    try {
+      const user_id = call.request.getUserId();
+
+      // @TODO create the Rapyd e_wallet here
```
Comment (netpoe) [inline wallet/src/service/controller.ts:33]: @Ktoxcon you can add the Rapyd calls here.

File: wallet/src/lib/rapyd/RapydClient.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  HttpMethods,
+  IRapydClient,
+  IRapydRequestParams,
+  RequestSignature,
+} from "./types";
+
+class RapydClient implements IRapydClient {
+  post<T>({ path, body }: IRapydRequestParams<T>): Promise<T> {
+    return new Promise<T>((resolve, reject) => {
+      const { BASE_URL, SALT_QTY, ACCESS_KEY, SECRET_KEY } = process.env;
```
Comment (netpoe) [inline wallet/src/lib/rapyd/RapydClient.ts:14]: I've never used it like this before. Nice!

File: wallet/src/lib/rapyd/RapydClient.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+import * as Crypto from "crypto-js";
+import moment from "moment";
+import {
+  HttpMethods,
+  IRapydClient,
+  IRapydRequestParams,
+  RequestSignature,
+} from "./types";
+
+class RapydClient implements IRapydClient {
```
Comment (netpoe) [inline wallet/src/lib/rapyd/RapydClient.ts:11]: This client is a great starting point to develop a Rapyd Typescript SDK in the future, which Rapyd developers would appreciate 🤔

Comment (Ktoxcon) [reply wallet/src/lib/rapyd/RapydClient.ts]: Yup, Developers using JS, TS and Node will love it.

**Review by Ktoxcon — APPROVED** [review]

File: wallet/src/__test__/client/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  beforeAll(async () => {
+    driver = await database.connect({ force: true });
+    dao = new WalletDAO(driver);
+    walletClient = new WalletClientGenerator(
+      `${process.env.IP_ADDRESS}:${process.env.HTTP_PORT}`
+    ).create();
+    rapydClient = new RapydClient();
+  });
+
+  test("success: creates a real Rapyd wallet per user and stores it in the database and in Rapyd", async () => {
```
Comment (netpoe) [inline wallet/src/__test__/client/index.test.ts:27]: @Ktoxcon I've begun testing the WalletClient, it's not fully working yet, but I'm getting close. 

I'm getting this error on the `GET wallet` endpoint: 

Screen Shot 2021-06-01 at 18 53 06

File: wallet/src/__test__/client/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+      const { id, ewallet_reference_id } =
+        await rapydClient.get<WalletObjectResponse>({
+          path: `/v1/user/${rapyd_ewallet_address}`,
+        });
+
+      expect(id).toEqual(rapyd_ewallet_address);
+      expect(ewallet_reference_id).toEqual(userId);
+    }
+  });
```
Comment (netpoe) [inline wallet/src/__test__/client/index.test.ts:62]: @Ktoxcon ya quedó 👍 

Testéemos así el Checkout y ya podemos concentrarnos en la parte de los comandos del Bot. 
Después de Checkout seguiría Send y después Withdraw.  

Qué opinas

File: wallet/src/__test__/client/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+          }
+
+          resolve(reply.getCheckoutPageUrl());
+        });
+      });
+
+    const checkout_page_urls = await Promise.all(users.map(topUpWallet));
+
+    for (const url of checkout_page_urls) {
+      expect(url).toBeTruthy();
```
Comment (netpoe) [inline wallet/src/__test__/client/index.test.ts:87]: @Ktoxcon 
Top Up Wallet is working 👍 

A few considerations: 

- The Bot will need to ask: Country Code, Currency Code and Amount
- Country Code and Currency Code can be options of Telegram Keyboard Markup Buttons

Screen Shot 2021-06-02 at 10 36 20

File: wallet/src/__test__/client/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        responseStatus: "accept",
+        pendingTransactionId,
+      });
+
+    expect(setTransferFromWalletResponseReply.amount).toEqual(requestAmount);
+    expect(setTransferFromWalletResponseReply.currency).toEqual(
+      requestCurrency
+    );
+    expect(setTransferFromWalletResponseReply.senderUserId).toEqual(users[0]);
+  });
```
Comment (netpoe) [inline wallet/src/__test__/client/index.test.ts:187]: @Ktoxcon 
Successfully transferring from wallet to wallet. 

Let's start experimenting with the Bot and how to handle these requests. 

Screen Shot 2021-06-02 at 15 07 09

Screen Shot 2021-06-02 at 15 06 53

**Pull request merged**